internal/graph/check.go

```diff
@@ -596,54 +655,60 @@ func (c *LocalChecker) produceUsersetDispatches(ctx context.Context, req *Resolv
 }
 
 // processDispatches returns a channel where the outcomes of the dispatched checks are sent, and begins sending messages to this channel.
-func (c *LocalChecker) processDispatches(ctx context.Context, limit uint32, dispatchChan chan dispatchMsg) chan checkOutcome {
+func (c *LocalChecker) processDispatches(ctx context.Context, limit uint32, dispatchChan chan dispatchMsg) (chan checkOutcome, chan error) {
 	outcomes := make(chan checkOutcome, limit)
 	dispatchPool := concurrency.NewPool(ctx, int(limit))
+	panicChan := make(chan error, 1)
 
 	go func() {
-		defer func() {
-			// We need to wait always to avoid a goroutine leak.
-			_ = dispatchPool.Wait()
-			close(outcomes)
-		}()
-
-		for {
-			select {
-			case <-ctx.Done():
-				return
-			case msg, ok := <-dispatchChan:
-				if !ok {
+		recoveredError := panics.Try(func() {
+			defer func() {
+				// We need to wait always to avoid a goroutine leak.
+				_ = dispatchPool.Wait()
+				close(outcomes)
+				close(panicChan)
+			}()
+			for {
+				select {
+				case <-ctx.Done():
 					return
-				}
-				if msg.err != nil {
-					concurrency.TrySendThroughChannel(ctx, checkOutcome{err: msg.err}, outcomes)
-					break // continue
-				}
-				if msg.shortCircuit {
-					resp := &ResolveCheckResponse{
-						Allowed: true,
+				case msg, ok := <-dispatchChan:
+					if !ok {
+						return
+					}
+					if msg.err != nil {
+						concurrency.TrySendThroughChannel(ctx, checkOutcome{err: msg.err}, outcomes)
+						break // continue
+					}
+					if msg.shortCircuit {
+						resp := &ResolveCheckResponse{
+							Allowed: true,
+						}
+						concurrency.TrySendThroughChannel(ctx, checkOutcome{resp: resp}, outcomes)
+						return
 					}
-					concurrency.TrySendThroughChannel(ctx, checkOutcome{resp: resp}, outcomes)
-					return
-				}
 
-				if msg.dispatchParams != nil {
-					dispatchPool.Go(func(ctx context.Context) error {
-						resp, err := c.dispatch(ctx, msg.dispatchParams.parentReq, msg.dispatchParams.tk)(ctx)
-						concurrency.TrySendThroughChannel(ctx, checkOutcome{resp: resp, err: err}, outcomes)
-						return nil
-					})
+					if msg.dispatchParams != nil {
+						dispatchPool.Go(func(ctx context.Context) error {
+							resp, err := c.dispatch(ctx, msg.dispatchParams.parentReq, msg.dispatchParams.tk)(ctx)
+							concurrency.TrySendThroughChannel(ctx, checkOutcome{resp: resp, err: err}, outcomes)
+							return nil
+						})
+					}
 				}
 			}
+		})
+		if recoveredError != nil {
+			panicChan <- fmt.Errorf("panic occurred: %v", recoveredError)
```

Comment:
If we leave panicChan as separate channel, be sure to use concurrency.TrySendThroughChannel